posts/2020-08-16-ocp-log-forwarding.adoc

```diff
@@ -0,0 +1,390 @@
+        name: fluentd
+   - name: elasticsearch-insecure 
+     type: "elasticsearch"  
+     endpoint: http://elasticsearch-insecure.offcluster.com:9200 
+     insecure: true
+   - name: fluentd-forward
+     type: "forward"
+     endpoint: https://splunk-fluentd-forward.offcluster.com:24224
+     secret:
+        name: secure-forward
+  pipelines: 
+   - name: container-logs 
+     inputSource: logs.app 
+     outputRefs: 
+     - elasticsearch
+     - elasticsearch-insecure
+     - fluentd-forward
+   - name: infra-logs
+     inputSource: logs.infra
+     outputRefs:
+     - elasticsearch
+   - name: audit-logs
+     inputSource: logs.audit
+     outputRefs:
+     - fluentd-forward
+```
++
+The example has three outputs defined, elasticsearch routing to internal Elasticsearch instance, elasticsearch routing to external insecure Elasticsearch cluster and forward routing to an instance of Fluentd. Each log type is defined under pipelines with its configured ouputs.
+
+. Create the instance:
++
+[source]
+----
+[root@ocp ~]# oc create -f log-forward-instance.yaml
+----
++
+
+. Annotate the ClusterLogging instance to enable the Log Forwarding API.
++
+[source]
+----
+[root@ocp ~]# oc annotate clusterlogging -n openshift-logging instance clusterlogging.openshift.io/logforwardingtechpreview=enabled
+----
++
+
+. Liberty application pods output logs in JSON format, therefore it is recommended to set Fluentd to parse the JSON fields from the message body. This feature is disabled by default. To enable it, set the cluster logging instance's *managementState* field from *"Managed"* to *"Unmanaged"*.
++
+```
+[root@ocp ~]# oc edit ClusterLogging instance
+
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogging"
+metadata:
+  name: "instance"
+
+....
+
+spec:
+  managementState: "Unmanaged"
+```
++
+Then, set the environment variable *MERGE_JSON_LOG* to *true*.
++
+[source]
+----
+[root@ocp ~]# oc set env ds/fluentd MERGE_JSON_LOG=true
+----
++
+
+. To check if the logs are being forwarded to the specified outputs, run the following command:
++
+[source]
+----
+[root@ocp ~]# oc -n openshift-logging get cm fluentd -o json | jq -r '.data."fluent.conf"' > fluentd-with-logfowarding.conf
+----
++
+This command gets ConfigMap configuration for Fluentd inside OpenShift Container Platform. Check if the outputs are defined inside the configuration file.
+
+* For example:
++
+```
+...
+<label @CONTAINER_LOGS>
+  <match **>
+    @type copy
+
+    <store>
+      @type relabel
+      @label @ELASTICSEARCH
+    </store>
+    <store>
+      @type relabel
+      @label @ELASTICSEARCH-INSECURE
+    </store>
+    <store>
+      @type relabel
+      @label @FLUENTD-FORWARD
+    </store>
+  </match>
+</label>
+...
+```
++
+
+
+== Internal and External Splunk configuration
+
+Using `forward` output, you can forward OpenShift Container Platform logs to Splunk using Fluentd forward protocol. You can setup Splunk inside your OpenShift Cluster or on your external machine.
```

Comment:
as discussed, refactor this section to make the external splunk scenario be your primary one and refer to the RH article for the splunk in OpenShift one.